citation/graphviz/data.py

```diff
@@ -1,171 +1,183 @@
-        start_year = datetime.strptime(filter_criteria.pop('date_published__gte'), '%Y-%m-%dT%H:%M:%SZ').year
-    if 'date_published__lte' in filter_criteria:
-        end_year = datetime.strptime(filter_criteria.pop('date_published__lte'), '%Y-%m-%dT%H:%M:%SZ').year
+        # fetching only filtered publication
+        self.publications = Publication.api.primary(**self.filter_criteria)
 
-    # fetching only filtered publication
-    primary_publications = Publication.api.primary(**filter_criteria)
+    def qualify(self, year_published):
+        if year_published and self.start_year <= year_published <= self.end_year:
+            return year_published
 
-    primary_pk = []
-    primary_pubs = []
-    for pub in primary_publications:
-        if pub.year_published is not None and start_year <= pub.year_published <= end_year:
-            primary_pk.append(pub.pk)
-            primary_pubs.append(pub)
+        return None
 
-    # fetches links that satisfies the given filter
-    links_candidates = primary_publications.filter(pk__in=primary_pk, citations__in=primary_pubs).values_list('pk',
-                                                                                                              'citations')
-    return links_candidates
 
+class NetworkVisualization(VisualizationBaseClass):
 
-def get_network_default_filter(group_by):
-    if group_by == NetworkGroupByType.SPONSOR.value:
-        return Publication.api.get_top_records(attribute='sponsors__name')
-    else:
-        return Publication.api.get_top_records(attribute='tags__name')
+    def __init__(self, filter_criteria=None, group_by=NetworkGroupByType.TAGS.value):
+        super().__init__(filter_criteria)
 
+        self.group_by = group_by
+        if group_by + '__name__in' in self.filter_criteria:
+            self.filter_value = self.filter_criteria[group_by + '__name__in']
+        else:
+            self.filter_value = self.get_default_group_by_filter()
+            self.filter_criteria[group_by + '__name__in'] = self.filter_value
 
-def generate_network_graph(filter_criteria, group_by=NetworkGroupByType.TAGS.value):
+    def get_default_group_by_filter(self):
+        if self.group_by == NetworkGroupByType.SPONSOR.value:
+            return Publication.api.get_top_records('sponsors__name', 5)
+        else:
+            return Publication.api.get_top_records('tags__name', 5)
+
+    def get_common_value(self, first, second):
+        """
+        :param first: list
+        :param second: list
+        :return: first common value found
+        """
+        for value in first:
+            if value in second:
+                return value
+
+        return None
+
+    # Generates links that will be used to form the network based on the provided filter criteria
+    def generate_link_candidates(self):
+        primary_pk = []
+        primary_pubs = []
+        for pub in self.publications:
+            if self.qualify(pub.year_published):
```

Comment:
It seems this `qualify` check could happen in the QuerySet itself and be done in SQL vs application code. If so, it should be a filter criteria instead. In a broader context we should chat about what this method is supposed to be doing and how to do it more efficiently.